frontend/src/lib/k8s/api/v2/webSocket.ts

```diff
@@ -1,224 +1,262 @@
 import { useEffect, useMemo } from 'react';
-import { findKubeconfigByClusterName, getUserIdFromLocalStorage } from '../../../../stateless';
-import { getToken } from '../../../auth';
-import { getCluster } from '../../../cluster';
+import { getUserIdFromLocalStorage } from '../../../../stateless';
+import { KubeObjectInterface } from '../../KubeObject';
 import { BASE_HTTP_URL } from './fetch';
-import { makeUrl } from './makeUrl';
+import { KubeList, KubeListUpdateEvent } from './KubeList';
 
+// Constants for WebSocket connection
 const BASE_WS_URL = BASE_HTTP_URL.replace('http', 'ws');
+/**
+ * Multiplexer endpoint for WebSocket connections
+ */
+const MULTIPLEXER_ENDPOINT = 'wsMultiplexer';
+
+// Message types for WebSocket communication
+interface WebSocketMessage {
+  /** Cluster ID */
+  clusterId: string;
+  /** API path */
+  path: string;
+  /** Query parameters */
+  query: string;
+  /** User ID */
+  userId: string;
+  /** Message type */
+  type: 'REQUEST' | 'CLOSE' | 'COMPLETE';
+}
 
 /**
- * Create new WebSocket connection to the backend
- *
- * @param url - WebSocket URL
- * @param options - Connection options
- *
- * @returns WebSocket connection
+ * Singleton class to manage WebSocket connections across the application
  */
-export async function openWebSocket<T>(
-  url: string,
-  {
-    protocols: moreProtocols = [],
-    type = 'binary',
-    cluster = getCluster() ?? '',
-    onMessage,
-  }: {
-    /**
-     * Any additional protocols to include in WebSocket connection
-     */
-    protocols?: string | string[];
-    /**
-     *
-     */
-    type: 'json' | 'binary';
-    /**
-     * Cluster name
-     */
-    cluster?: string;
-    /**
-     * Message callback
-     */
-    onMessage: (data: T) => void;
-  }
-) {
-  const path = [url];
-  const protocols = ['base64.binary.k8s.io', ...(moreProtocols ?? [])];
-
-  const token = getToken(cluster);
-  if (token) {
-    const encodedToken = btoa(token).replace(/=/g, '');
-    protocols.push(`base64url.bearer.authorization.k8s.io.${encodedToken}`);
+class WebSocketManager {
+  private static instance: WebSocketManager;
```

Comment:
Instead of singleton pattern I'd suggest using a plain object for this.

```ts
const WebSocketManager = {
  socket: WebSocket | null = null,
  async connect() {
    // ...
  }
}